### PR TITLE
Yocto dev init

### PR DESCRIPTION
**Pull request by DAA2310**

### CODE REVIEW

File: yocto-dev/build_conf/bblayers.conf [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	${BSPDIR}/sources/meta-compulab-bsp/meta-desktop \
+	${BSPDIR}/sources/meta-compulab-bsp/meta-multimedia \
+	${BSPDIR}/sources/meta-compulab-bsp/meta-graphics \
+	${BSPDIR}/sources/meta-compulab-bsp/meta-utils \
+	${BSPDIR}/sources/meta-compulab-bsp/meta-bsp \
+	${BSPDIR}/sources/meta-compulab-bsp/meta-devtool \
+	${BSPDIR}/sources/meta-compulab-uefi \
+"
+
+# additional layers for Ice Cream
```
Comment (sarika93) [inline yocto-dev/build_conf/bblayers.conf:46]: Remove code name from comment. Mention what we actually need these ROS layers for e.g. the following dependencies needed to run Kimera-VIO can be found in <layer>.

File: yocto-dev/build_conf/local.conf [hunk 2]
```diff
@@ -0,0 +1,41 @@
+MACHINE ??= 'ucm-imx93'
```
Comment (sarika93) [inline yocto-dev/build_conf/local.conf:1]: Was the template for this taken from somewhere or did you write this from scratch? If the former is the case, you should add a reference to the original (repo link etc).

File: yocto-dev/build_conf/local.conf [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+LICENSE_FLAGS_ACCEPTED = "commercial"
+
+CORE_IMAGE_EXTRA_INSTALL += " cl-uboot cl-deploy cl-growfs-rootfs "
+CORE_IMAGE_EXTRA_INSTALL += " libubootenv-bin u-boot-compulab-env "
+CORE_IMAGE_EXTRA_INSTALL += " memtester htop iotop tmux iperf3 "
+CORE_IMAGE_EXTRA_INSTALL += " libgpiod-tools bluez5 can-utils minicom eeprom-util usbutils i2c-tools "
+CORE_IMAGE_EXTRA_INSTALL += " wireless-tools dhcpcd ifupdown wpa-supplicant ntp ntpdate "
+CORE_IMAGE_EXTRA_INSTALL += " networkmanager modemmanager "
+
+# additional recipes for Ice Cream
```
Comment (sarika93) [inline yocto-dev/build_conf/local.conf:34]: Same comment about code name as before. Also be more specific about how this is being used. Example: Recipes for Kimera-VIO and its dependencies.

File: yocto-dev/meta-custom/README.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+Kirkstone Compatible Layer with recipes for Ice Cream Project:
```
Comment (sarika93) [inline yocto-dev/meta-custom/README.md:1]: Remove code name. Be more specific about what these are being used for.

File: yocto-dev/meta-custom/conf/layer.conf [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# We have a conf and classes directory, add to BBPATH
```
Comment (sarika93) [inline yocto-dev/meta-custom/conf/layer.conf:1]: This comment is not useful. Please add more context.

File: yocto-dev/scripts/gtsam_som_test/test_gtsam.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
+/* ----------------------------------------------------------------------------
```
Comment (sarika93) [inline yocto-dev/scripts/gtsam_som_test/test_gtsam.cpp:1]: Do you think this is useful to include this directory in this PR?

File: yocto-dev/meta-custom/recipes-custom/dbow2/dbow2_1.1.bb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+SRC_URI = "git://github.com/dorian3d/DBoW2.git;protocol=https"
+SRCREV = "3924753db6145f12618e7de09b7e6b258db93c6e"
+
+# Dependencies
+DEPENDS = "cmake opencv"
+
+# Specify the directory to fetch the source code
+S = "${WORKDIR}/git"
+
+# Inherit classes
```
Comment (sarika93) [inline yocto-dev/meta-custom/recipes-custom/dbow2/dbow2_1.1.bb:17]: What classes?

File: yocto-dev/meta-custom/recipes-custom/dbow2/dbow2_1.1.bb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+inherit cmake
+
+FILES:${PN}-staticdev += "${libdir}/libDBoW2.so"
+
+ALLOW_EMPTY:${PN} = "1"
```
Comment (sarika93) [inline yocto-dev/meta-custom/recipes-custom/dbow2/dbow2_1.1.bb:22]: Please add a comment for why this is needed.

**Review by sarika93 — CHANGES_REQUESTED** [review]
If you are adding recipes you've taken from elsewhere and made some small modifications to, please include a reference to the original source code. 

There is also not enough context provided about why things are being done. Please include this where necessary but try to be as brief and to the point as possible.

File: yocto-dev/scripts/gtsam_som_test/test_gtsam.cpp (hunk 6 above)
Comment (DAA2310) [reply yocto-dev/scripts/gtsam_som_test/test_gtsam.cpp:1]: Not really, will remove and add later.

File: yocto-dev/README.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+- Standard configuration files for each development machine
+- Miscellaneous image related scripts 
+
+**How to setup?**
+
+- Configure environment prerequisites for Yocto. 
+    - Install on [local environment](https://docs.yoctoproject.org/brief-yoctoprojectqs/index.html#yocto-project-quick-build) OR
+    - Utilize Compulab's [Docker containers](https://github.com/compulab-yokneam/yocker). This should be selected for Yocto version - Kirkstone.
+- Import and set up Compulab's BSP for the IMX9
+    - Follow [instructions](https://github.com/compulab-yokneam/meta-bsp-imx9/tree/kirkstone-2.2.0) up to _**"Setup build environment"**_.
```
[the remark this answers is not in the document]
Comment (DAA2310) [reply yocto-dev/README.md:14]: Modified the instruction to make it clear that it's all the instructions up until this final line for "Setup build env"

File: yocto-dev/build_conf/local.conf (hunk 2 above)
Comment (DAA2310) [reply yocto-dev/build_conf/local.conf:1]: This file is auto-generated when sourcing the Compulab's `compulab-setup-env` script for the first time as part of their README. This is appended to and we need to overwrite a newly generated version of this file on a fresh system, and similarly the `bblayers.conf`, with the appended Kimera related lines for our image. The ideal workflow to follow would be to implement a setup script to do this for us, as well as any other steps which can be automated.

File: yocto-dev/meta-custom/conf/layer.conf (hunk 5 above)
Comment (DAA2310) [reply yocto-dev/meta-custom/conf/layer.conf:1]: This entire file is auto-generated when creating a layer from scratch using `bitbake-layers` functionality. Not sure whether it's still worth commenting.

File: yocto-dev/README.md [hunk 10]
```diff
@@ -1,30 +1,35 @@
...
+# Yocto Development
 
-This folder includes all relevant artifacts for setting up an independent Yocto development environment for project Ice Cream:
+This folder includes all relevant artifacts for setting up an independent Yocto development environment for creating our own custom images:
 - Custom Yocto layer for additional/modified recipes for IMX image
 - Standard configuration files for each development machine
 - Miscellaneous image related scripts 
 
-**How to setup?**
+#### Setting up Development Environment
```
[the remark this answers is not in the document]
Comment (sarika93) [reply yocto-dev/README.md:8]: Also add a sub-section here with the minimum system requirements for compiling Yocto (memory, disk space etc).

File: yocto-dev/meta-custom/README.md [hunk 11]
```diff
@@ -1,4 +1,5 @@
-Kirkstone Compatible Layer with recipes for Ice Cream Project:
+#### Kirkstone-Compatible Layer for Kimera-VIO
```
Comment (sarika93) [inline yocto-dev/meta-custom/README.md:1]: Change to Heading 1.

**Review by sarika93 — CHANGES_REQUESTED** [review]
Some minor readme changes to make. Recipe comments look fine to me.

File: yocto-dev/README.md [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    - Verify that the root of the directory created above includes a `build` and `sources` folder. 
+- In the build folder, substitute _bblayers.conf_ and _local.conf_ for their respective files in this repo's `yocto_dev/build_conf` folder. 
+    - _bblayers.conf_ - Defines the layers referenced by bitbake 
+    - _local.conf_ - Defines the recipes referenced by bitbake
+    See [Yocto Project Concepts](https://docs.yoctoproject.org/4.0.20/singleindex.html#yocto-project-concepts) for more information on bitbake, these config files and more.
+- In the sources folder, we need to import additional layers that are build prerequisites:
+    - Clone this **indoor_drone_experimentation** repo to include dependencies and custom recipe for Kimera-VIO.
+    - Clone the [meta-ros](https://github.com/ros/meta-ros/tree/kirkstone) (kirkstone branch) for GTSAM recipe dependencies.
+
+## Building a UCM-iMX93 image on your Machine
```
Comment (sarika93) [inline yocto-dev/README.md:29]: Leaving this as a question for you, rather than me telling you the answer. Is this the most useful place to include this wrt flow of the readme? How did including this here affect the information below?

File: yocto-dev/build_conf/local.conf (hunk 2 above)
Comment (sarika93) [reply yocto-dev/build_conf/local.conf:1]: If we overwrite this every time we set up our environment on a new machine, then it should not be included in this repository at all because every time someone does setup, it will show up as a change. Please add a .gitignore file to this directory and include this file there.

File: yocto-dev/meta-custom/conf/layer.conf (hunk 5 above)
Comment (sarika93) [reply yocto-dev/meta-custom/conf/layer.conf:1]: See above comment about `yocto-dev/build_conf/local.conf`. If this is not the same situation, then add a one line readme to this directory stating what what this file is and that it is auto-generated.

File: yocto-dev/README.md [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+Minimum system requirements:
+- 4 CPU Cores
+- 8GB RAM
+- 90GB Available Storage
```
Comment (sarika93) [inline yocto-dev/README.md:34]: Compulab [recommends](https://mediawiki.compulab.com/w/index.php?title=Yocto:_Setting_Up_Build_Environment) the following:

> 4 CPU cores, 64GB RAM, and 500GB of free disk space

The recommendation you included here seems like cutting it very close? Even NXP recommends a minimum 120GB.

**Review by sarika93 — CHANGES_REQUESTED** [review]

File: yocto-dev/README.md [hunk 14]
```diff
@@ -0,0 +1,41 @@
+# Yocto Development
```
Comment (sarika93) [inline yocto-dev/README.md:1]: Your readme is missing a section for how to actually use the image to run Kimera on the UCM-i.MX93.

File: yocto-dev/README.md (hunk 13 above)
Comment (DAA2310) [reply yocto-dev/README.md:34]: I was using the minimum from Yocto's documentation but I will take NXP's minimum storage requirement. Minimum RAM only suggests a slower build, whereas insufficient storage space can prevent builds entirely.

File: yocto-dev/build_conf/local.conf (hunk 2 above)
Comment (DAA2310) [reply yocto-dev/build_conf/local.conf:1]: The files which are autogenerated are actually located in the `{BUILD_FOLDER_NAME}/conf` folder in the root of Compulab's setup directory.

Our `build_conf` folder does not mimic the structure and therefore will not show up as a change. This is simply a folder to store the configs we require, hence the need to overwrite the ones autogenerated.

File: yocto-dev/README.md [hunk 15]
```diff
@@ -26,16 +33,34 @@ This folder includes all relevant artifacts for setting up an independent Yocto
...
 
-Minimum system requirements:
-- 4 CPU Cores
-- 8GB RAM
-- 90GB Available Storage
+*Note:* Every time you restart your computer OR enter the Docker dev container, you need to source the build folder: 
+```
+source compulab-setup-env {BUILD_FOLDER_NAME}
+```
+
```
Comment (sarika93) [inline yocto-dev/README.md:45]: Nit: A lot of unnecessary whitespace here.

**Review by sarika93 — DISMISSED** [review]
One minor comment, otherwise LGTM

**Review by sarika93 — APPROVED** [review]
